File: docs/assets/js/feed.js
/**
 * Deterministic congestion feed simulator.
 *
 * Browser counterpart of `src/pulseroute/feed.py`. In production this layer is
 * replaced by real telemetry (gate scan rates, Wi-Fi association counts, CV
 * people-counting); consumers only depend on the abstract shape it returns:
 * `{ zone: congestionFactor }`.
 */

export const ZONES = ["ne", "se", "sw", "nw"];

/**
 * Scripted matchday timeline (kickoff at 0', halftime ~45', full-time ~100').
 * Values are congestion multipliers: 0 = free-flowing, 1.0 doubles traversal time.
 * Kept sorted by minute so lookup is a binary search rather than a re-sort.
 */
const TIMELINE = [
  { minute: 0, factors: { ne: 0.9, se: 0.3, sw: 0.2, nw: 0.7 },
    note: "Pre-kickoff arrival surge at north/east gates." },
  { minute: 45, factors: { ne: 0.4, se: 1.6, sw: 1.4, nw: 0.5 },
    note: "Halftime: concession rush jams south concourse." },
  { minute: 100, factors: { ne: 1.8, se: 1.2, sw: 1.1, nw: 1.7 },
    note: "Full-time egress: heavy load on all exits." },
];

/** Index of the last entry whose minute is <= `minute` (binary search). */
function indexAtOrBefore(entries, minute) {
  let lo = 0;
  let hi = entries.length - 1;
  let found = 0;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    if (entries[mid].minute <= minute) {
      found = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  return found;
}

export class FeedSimulator {
  constructor(timeline = TIMELINE) {
    if (!timeline.length) throw new Error("FeedSimulator requires a non-empty timeline.");
    this.timeline = [...timeline].sort((a, b) => a.minute - b.minute);
  }

  /** The most recent scripted snapshot at or before `minute`. */
  snapshot(minute) {
    const entry = this.timeline[indexAtOrBefore(this.timeline, minute)];
    return { minute, factors: { ...entry.factors }, note: entry.note };
  }
}

File: src/pulseroute/feed.py
"""Deterministic congestion feed simulator.

In production this layer would be replaced by real telemetry (gate scan rates,
Wi-Fi association counts, CV people-counting). We ship a reproducible simulator
so the demo and tests are fully offline and deterministic. The rest of the system
consumes only the abstract interface: a dict of {zone: congestion_factor}.
"""
from __future__ import annotations

import json
from bisect import bisect_right
from dataclasses import dataclass
from pathlib import Path

ZONES = ("ne", "se", "sw", "nw")


@dataclass(frozen=True, slots=True)
class CongestionSnapshot:
    minute: int
    factors: dict[str, float]     # zone -> multiplier delta (0.0 = free-flowing)
    note: str

    def risk_zones(self, threshold: float = 0.8) -> list[str]:
        return sorted(
            (z for z, f in self.factors.items() if f >= threshold),
            key=lambda z: self.factors[z],
            reverse=True,
        )


# A scripted matchday timeline (kickoff at minute 0, halftime ~45, full-time ~100).
# Values are congestion multipliers: 0.0 free, 1.0 = doubles traversal time.
_TIMELINE = {
    0:   ({"ne": 0.9, "se": 0.3, "sw": 0.2, "nw": 0.7}, "Pre-kickoff arrival surge at north/east gates."),
    45:  ({"ne": 0.4, "se": 1.6, "sw": 1.4, "nw": 0.5}, "Halftime: concession rush jams south concourse."),
    100: ({"ne": 1.8, "se": 1.2, "sw": 1.1, "nw": 1.7}, "Full-time egress: heavy load on all exits."),
}


class FeedSimulator:
    """Serves the scripted snapshot in effect at a given matchday minute.

    The timeline keys are sorted **once** at construction, so each lookup is an
    ``O(log n)`` binary search rather than re-sorting the timeline per call.
    """

    def __init__(self, timeline: dict[int, tuple[dict[str, float], str]] | None = None):
        self._timeline = timeline or _TIMELINE
        self._keys = sorted(self._timeline)
        if not self._keys:
            raise ValueError("FeedSimulator requires a non-empty timeline.")

    def snapshot(self, minute: int) -> CongestionSnapshot:
        """Return the most recent scripted snapshot at or before `minute`."""
        idx = bisect_right(self._keys, minute) - 1
        key = self._keys[max(idx, 0)]
        factors, note = self._timeline[key]
        return CongestionSnapshot(minute=minute, factors=dict(factors), note=note)

    @classmethod
    def from_json(cls, path: str | Path) -> FeedSimulator:
        raw = json.loads(Path(path).read_text(encoding="utf-8"))
        timeline = {int(k): (v["factors"], v["note"]) for k, v in raw.items()}
        return cls(timeline)
